### PR TITLE
pid1: enqueue auto-restart job for the unit itself as JOB_START and for dependending units as JOB_RESTART

### LINKED ISSUE #26903: OnSuccess is triggered when using Restart=on-failure
### systemd version the issue has been seen with

253.1 / 252.6

### Used distribution

Arch / Debian 12

### Linux kernel version used

6.2.7-zen1-1-zen / 6.1.0-6-amd64

### CPU architectures issue was seen on

x86_64

### Component

systemd

### Expected behaviour you didn't see

`OnSuccess` dependencies are triggered after the unit completed successfully.

### Unexpected behaviour you saw

`OnSuccess` dependencies are also triggered while the unit is restarting, when `Restart=on-failure` is set.

### Steps to reproduce the problem

Use two dummy units

```
# /etc/systemd/system/myservice.service
[Unit]
OnSuccess=mysuccess.service
StartLimitBurst=2
StartLimitIntervalSec=1h


[Service]
Type=oneshot
ExecStart=/usr/bin/false
Restart=on-failure
```

```
# /etc/systemd/system/mysuccess.service
[Service]
Type=oneshot
ExecStart=/usr/bin/true
```

When starting `myservice.service`, this will also trigger `mysuccess.service` multiple times.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
Mar 20 15:49:21 periculum systemd[1]: Starting myservice.service...
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Child 4446 belongs to myservice.service.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Main process exited, code=exited, status=1/FAILURE
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Failed with result 'exit-code'.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Service will restart (restart setting)
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Changed start -> failed
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Job 3357 myservice.service/start finished, result=failed
Mar 20 15:49:21 periculum systemd[1]: Failed to start myservice.service.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Unit entered failed state.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Consumed 1ms CPU time.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Changed failed -> auto-restart
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Service RestartSec=100ms expired, scheduling restart.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Trying to enqueue job myservice.service/restart/replace
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Installed new job myservice.service/restart as 3481
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Enqueued job myservice.service/restart as 3481
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Scheduled restart job, restart counter is at 1.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Changed auto-restart -> dead
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Job 3481 myservice.service/restart finished, result=done
Mar 20 15:49:21 periculum systemd[1]: Stopped myservice.service.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Converting job myservice.service/restart -> myservice.service/start
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Consumed 1ms CPU time.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Triggering OnSuccess= dependencies.
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Trying to enqueue job mysuccess.service/start/fail
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Installed new job mysuccess.service/start as 3605
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Enqueued job mysuccess.service/start as 3605
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Triggering OnSuccess= dependencies done (1 job).
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Will spawn child (service_enter_start): /usr/bin/false
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Passing 0 fds to service
Mar 20 15:49:21 periculum systemd[1]: myservice.service: About to execute /usr/bin/false
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Forked /usr/bin/false as 4448
Mar 20 15:49:21 periculum (false)[4448]: myservice.service: Executing: /usr/bin/false
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Changed dead -> start
Mar 20 15:49:21 periculum systemd[1]: Starting myservice.service...
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Will spawn child (service_enter_start): /usr/bin/true
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Passing 0 fds to service
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: About to execute /usr/bin/true
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Forked /usr/bin/true as 4449
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Changed dead -> start
Mar 20 15:49:21 periculum systemd[1]: Starting mysuccess.service...
Mar 20 15:49:21 periculum (true)[4449]: mysuccess.service: Executing: /usr/bin/true
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Child 4448 belongs to myservice.service.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Main process exited, code=exited, status=1/FAILURE
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Failed with result 'exit-code'.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Service will restart (restart setting)
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Changed start -> failed
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Job 3481 myservice.service/start finished, result=failed
Mar 20 15:49:21 periculum systemd[1]: Failed to start myservice.service.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Unit entered failed state.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Consumed 1ms CPU time.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Changed failed -> auto-restart
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Child 4449 belongs to mysuccess.service.
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Main process exited, code=exited, status=0/SUCCESS (success)
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Deactivated successfully.
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Service will not restart (restart setting)
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Changed start -> dead
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Job 3605 mysuccess.service/start finished, result=done
Mar 20 15:49:21 periculum systemd[1]: Finished mysuccess.service.
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Consumed 1ms CPU time.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Control group is empty.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Service RestartSec=100ms expired, scheduling restart.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Trying to enqueue job myservice.service/restart/replace
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Installed new job myservice.service/restart as 3729
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Enqueued job myservice.service/restart as 3729
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Scheduled restart job, restart counter is at 2.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Changed auto-restart -> dead
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Job 3729 myservice.service/restart finished, result=done
Mar 20 15:49:21 periculum systemd[1]: Stopped myservice.service.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Converting job myservice.service/restart -> myservice.service/start
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Consumed 1ms CPU time.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Triggering OnSuccess= dependencies.
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Trying to enqueue job mysuccess.service/start/fail
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Installed new job mysuccess.service/start as 3853
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Enqueued job mysuccess.service/start as 3853
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Triggering OnSuccess= dependencies done (1 job).
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Start request repeated too quickly.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Failed with result 'exit-code'.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Service restart not allowed.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Changed dead -> failed
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Job 3729 myservice.service/start finished, result=failed
Mar 20 15:49:21 periculum systemd[1]: Failed to start myservice.service.
Mar 20 15:49:21 periculum systemd[1]: myservice.service: Unit entered failed state.
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Will spawn child (service_enter_start): /usr/bin/true
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Passing 0 fds to service
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: About to execute /usr/bin/true
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Forked /usr/bin/true as 4450
Mar 20 15:49:21 periculum (true)[4450]: mysuccess.service: Executing: /usr/bin/true
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Changed dead -> start
Mar 20 15:49:21 periculum systemd[1]: Starting mysuccess.service...
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Child 4450 belongs to mysuccess.service.
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Main process exited, code=exited, status=0/SUCCESS (success)
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Deactivated successfully.
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Service will not restart (restart setting)
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Changed start -> dead
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Job 3853 mysuccess.service/start finished, result=done
Mar 20 15:49:21 periculum systemd[1]: Finished mysuccess.service.
Mar 20 15:49:21 periculum systemd[1]: mysuccess.service: Consumed 2ms CPU time.
```
```

### PR DESCRIPTION
**Pull request by poettering**
A fix for #27722.

(Not tested thoroughly yet, hence 'dont-merge')

Fixes #26903

### CODE REVIEW

File: src/core/service.c [hunk 1]
```diff
@@ -2510,17 +2510,15 @@ static void service_enter_restart(Service *s) {
...
         }
 
-        /* Any units that are bound to this service must also be
-         * restarted. We use JOB_RESTART (instead of the more obvious
-         * JOB_START) here so that those dependency jobs will be added
-         * as well. */
-        r = manager_add_job(UNIT(s)->manager, JOB_RESTART, UNIT(s), JOB_REPLACE, NULL, &error, NULL);
+        /* Any units that are bound to this service must also be restarted. We use JOB_RESTART (instead of
+         * the more obvious JOB_START) here so that those dependency jobs will be added as well. */
+        r = manager_add_job(UNIT(s)->manager, JOB_RESTART, UNIT(s), JOB_RESTART_DEPENDENCIES, NULL, &error, NULL);
```
Comment (YHNdnzj) [inline src/core/service.c]:
```suggestion
        r = manager_add_job(UNIT(s)->manager, JOB_START, UNIT(s), JOB_RESTART_DEPENDENCIES, NULL, &error, NULL);
```

File: src/core/service.c [hunk 2]
```diff
@@ -2536,10 +2534,8 @@ static void service_enter_restart(Service *s) {
         /* Notify clients about changed restart counter */
         unit_add_to_dbus_queue(UNIT(s));
 
-        /* Note that we stay in the SERVICE_AUTO_RESTART state here,
-         * it will be canceled as part of the service_stop() call that
-         * is executed as part of JOB_RESTART. */
-
+        /* Note that we stay in the SERVICE_AUTO_RESTART state here, it will be canceled as part of the
+         * service_stop() call that is executed as part of JOB_RESTART. */
```
Comment (YHNdnzj) [inline src/core/service.c]: This should be updated/removed?

**Comment by poettering** [comment]
Reworked it a bit. Seems to work great. Still need to put together a test for it.

**Comment by poettering** [comment]
add testcase now. ptal

**Review by bluca — APPROVED** [review]

**Comment by Cornelicorn** [comment]
@bluca Would it be possible to backport this fix into the stable releases?

**Pull request merged**